Sources/MessagingPush/PushHandling/iOSPushEventListener.swift

```diff
@@ -44,12 +44,19 @@ class IOSPushEventListener: PushEventHandler {
 
         logger.debug("Push came from CIO. Handle the didReceive event on behalf of the customer.")
 
-        if pushAction.didClickOnPush {
-            pushClickHandler.pushClicked(push)
-        }
-
-        // call the completion handler so the customer does not need to.
-        completionHandler()
+        // Forward event to other push click handlers so they can receive a callback about this push event.
+        pushEventHandlerProxy.onPushAction(pushAction, completionHandler: {
+            // When this block of code executes, the customer is done processing the push event.
+
+            // We do not open deep link until after customer is done processing the push event in case the deep link would leave the app.
+            // We want to make sure the customer has a chance to process the push event before leaving the app.
+            if pushAction.didClickOnPush {
+                self.pushClickHandler.pushClicked(push)
+            }
+
+            // call the completion handler, indicating to the OS that we are done processing the push.
+            completionHandler()
```

Comment:
while I can tell how this fixes the existing issue, I see it possibly introducing another issue.
What if the customer didn't call the completionHandler we are passing here `pushEventHandlerProxy.onPushAction` ? neither our code or the customer code will run

Reply:
I wish that was not the case, but that is correct. 

I called this out and provided more details about this in the PR description. 

The only way that I can think of to avoid this scenario is to call the 3rd party callback functions and not wait for them to call the async completion handler. But, that runs the risk of customer's push processing getting terminated early by the OS which I think is worse. 
 
Terminating customer's processing a push + Apple's docs saying it's required to call completion handler when done processing a push, I do think this is the best implementation we have available.

Reply:
What's the problem with doing this
```
 if pushAction.didClickOnPush {
       self.pushClickHandler.pushClicked(push)
}

pushEventHandlerProxy.onPushAction(pushAction, completionHandler: {

            // call the completion handler, indicating to the OS that we are done processing the push.
            completionHandler()
}
 ```
 
 At this point, our click handler guaranteed to run, no?
 If there is a problem on the client side that doesn't call the `completionHandler` then it is a bug in their code and has nothing to do with our implementation, no?

Reply:
> but i believe the callback onPushAction in only called, when all the other delegates have called the completionHandler.

Which is yet another concern. I will respond to your suggestion. At this point my gut feeling is that I need to understand the problem we are solving better. Otherwise, I am feeling like I am contributing to work around a problem that's not fully clear to me

Reply:
In a call today, we made the decision to move forward with de-coupling `opened` metrics tracking and opening deep links. That way we can make this feature more reliable, in case someone doesn't call the async completion handler. 

[I created a sub-issue for that change](https://linear.app/customerio/issue/MBL-154/track-opened-push-metrics-3rd-party-push-callback-functions-get-called) and will open another feature for that change onto this stack. 

I believe that this means this conversation can be resolved.

Reply:
[New PR](https://github.com/customerio/customerio-ios/pull/590) introduced in this stack to cover this.